midterm_project_Nechita_Zaganeanu/shopping_list.py

```diff
@@ -0,0 +1,268 @@
+        name = self.name
+        star_line = '*' * (len(name)+4) + '\n'
+        string = star_line + '  '+ name + '\n'
+        string += star_line + '\n'     
+        for index, (keys, values) in enumerate(self.items(), 1):
+            string += f'  {index}. {keys}: {values} \n'
+        string += '\n'
+        string += star_line
+        return (string)
+
+#taken from https://stackoverflow.com/questions/19022868/how-to-make-dictionary-read-only-in-python
+class ReadOnlyDict(dict):
+    
+    def __readonly__(self, *args, **kwargs):
+        raise RuntimeError("Cannot modify Recipe")
+    
+    __setitem__ = __readonly__
+    __delitem__ = __readonly__
+    pop = __readonly__
+    popitem = __readonly__
+    clear = __readonly__
+    update = __readonly__
+    setdefault = __readonly__
+    del __readonly__
+    
+class Recipe(PrettyPrinter):
+    
+    def __init__(self, name, dictionary_of_ingredients):
+        self._name = name
+        self._dictionary_of_ingredients = ReadOnlyDict(dictionary_of_ingredients)
+    
+    @property
+    def name(self):
+        return self._name
+
+    def __getitem__(self, ingredient):
+        return self._dictionary_of_ingredients[ingredient]
+    
+    def __contains__(self, ingredient):
+        if ingredient in self._dictionary_of_ingredients:
+            return True
+        else:
+            return False
+    def get(self, key):
+        return self._dictionary_of_ingredients[key]
+    
+    def __eq__(self, other):
+        if isinstance(other, Recipe):
+            return self._dictionary_of_ingredients == other._dictionary_of_ingredients
+        return False
+    
+    def __ne__(self, other):
+        return not self.__eq__(other)
+       
+    def __iter__(self):
+        return iter(self._dictionary_of_ingredients)       
+        
+    def __len__(self):
+        return len(self._dictionary_of_ingredients)
+    
+    def items(self):
+        return self._dictionary_of_ingredients.items()
+    
+    def keys(self):
+        return self._dictionary_of_ingredients.keys()
+    
+    def values(self):
+        return self._dictionary_of_ingredients.values()
+       
+    def __str__(self):
+        return super().__str__()
+                    
+class RecipesBox():
+   
+    def __init__ (self, name, recipes_list):
+        self.name=name
+        self.recipes_list = recipes_list
+    
+    def __iter__(self):
+        return iter(self.recipes_list)
+    
+    def insert (self, index, recipe):
+        return self.recipes_list.insert(index, recipe)
+    
+    def __getitem__(self, index):
+        return self.recipes_list[index]
+    
+    def __setitem__(self, index, value):
+        self.recipes_list[index]=value
+    
+    def __delitem__(self, index):
+        del self.recipes_list[index]
+        
+    def __len__(self):
+        return len(self.recipes_list)
+    
+    def pick(self, recipe_to_extract=None):
+        if recipe_to_extract:
+            for recipe in self.recipes_list:

```

Comment:
aici poti obtine indexul unui recipe din recipes_list folosind `self.recipes_list.index(recipe_to_xtract)`

https://docs.python.org/3/tutorial/datastructures.html

Reply:
recipe_to_extract este un nume, care este un atribut al obiectului de tip recipe din lista. deci self.recipes_list.index(recipe_to_extract) nu va fi gasit, deoarece recipe_to_extract nu va fi gasit in lista de retete.